### PR TITLE
[SPARK-15165][SPARK-15205][SQL] Introduce place holder for comments in generated code

### PR DESCRIPTION
**Pull request by sarutak**
## What changes were proposed in this pull request?

This PR introduce place holder for comment in generated code and the purpose  is same for #12939 but much safer.

Generated code to be compiled doesn't include actual comments but includes place holder instead.

Place holders in generated code will be replaced with actual comments only at the time of  logging.

Also, this PR can resolve SPARK-15205.
## How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by davies** [comment]
@sarutak It's expected to compile twice on two different queries, it does not worth to optimize this corner case (ideally it should generate different source code even without comments).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 1]
```diff
@@ -97,15 +97,20 @@ abstract class Expression extends TreeNode[Expression] {
     ctx.subExprEliminationExprs.get(this).map { subExprState =>
       // This expression is repeated which means that the code to evaluate it has already been added
       // as a function before. In that case, we just re-use it.
-      val code = s"/* ${toCommentSafeString(this.toString)} */"
-      ExprCode(code, subExprState.isNull, subExprState.value)
+      val placeHolder = s"/*{${ctx.freshName("comment_placeholder")}}*/"
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: Why add extra `{}` here? we could also make `comment_placeholder` shorter

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 2]
```diff
@@ -706,6 +712,20 @@ class CodegenContext {
...
+   * Add a pair of a place holder and a corresponding comment
+   */
+  def addCommentEntry(placeHolder: String, comment: String): Unit = {
+    placeHolderToCommentMap += (placeHolder -> comment)
+  }
+
+  /**
+   * Copy an immutable map of the pair of a place holder and a corresponding comment
+   */
+  def copyPlaceHolderToCommentMap(): immutable.Map[String, String] = {
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: CodegenContext is not re-used, do we still need this copy?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 3]
```diff
@@ -200,6 +201,11 @@ class CodegenContext {
   var freshNamePrefix = ""
 
   /**
+   * The map from a place holder to a corresponding comment
+   */
+  private val placeHolderToCommentMap = new mutable.HashMap[String, String]
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: placeHolderToComments

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 4]
```diff
@@ -717,6 +737,23 @@ abstract class GeneratedClass {
...
 /**
+ * A wrapper for the source code to be compiled by [[CodeGenerator]].
+ */
+class CodeAndComment(val body: String, val comment: Map[String, String]) extends Serializable {
+  override def equals(that: Any): Boolean = {
+    if (!that.isInstanceOf[CodeAndComment]) {
+      return false
+    }
+    val thatSourceCode = that.asInstanceOf[CodeAndComment]
+    if (thatSourceCode eq null) return false
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: We could use match-case to simplify this

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateMutableProjection.scala [hunk 5]
```diff
@@ -133,7 +133,9 @@ object GenerateMutableProjection extends CodeGenerator[Seq[Expression], MutableP
       }
     """
 
-    logDebug(s"code for ${expressions.mkString(",")}:\n${CodeFormatter.format(code)}")
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateMutableProjection.scala]: this line does not need to change

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGenerationSuite.scala [hunk 6]
```diff
@@ -177,6 +177,55 @@ class CodeGenerationSuite extends SparkFunSuite with ExpressionEvalHelper {
     assert(internalRow === internalRow2)
   }
 
+  test("reuse cached code") {
```
Comment (davies) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CodeGenerationSuite.scala]: remove this test case

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala [hunk 7]
```diff
@@ -333,7 +333,11 @@ case class WholeStageCodegenExec(child: SparkPlan) extends UnaryExecNode with Co
       """.trim
 
     // try to compile, helpful for debug
-    val cleanedSource = CodeFormatter.stripExtraNewLines(source)
+    val cleanedSource =
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala]: Could you also update the comment for plan? see L310

**Comment by davies** [comment]
These place holders are good for safety, could you update the PR title and description?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 1 above)
Comment (sarutak) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: Actually, `}` is really needed.
If we have place holders like `comment_placeholder1` and `comment_placeholder12`, and the actual comment `Hello` corresponds to `comment_placeholder1`, `comment_placeholder1` will be replaced with `Hello` but also `comment_placeholder12` will be replaced with `Hello2`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateMutableProjection.scala (hunk 5 above)
Comment (sarutak) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateMutableProjection.scala]: Sorry It's not clear for me why this line does not need to change.

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateMutableProjection.scala]: format() could accept CodeAndComment now, and code is CodeAndComment

Comment (sarutak) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateMutableProjection.scala]: I see. You mean we don't need change the line to `logDebug(s"code for ${expressions.mkString(",")}:\n$formatted")`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 8]
```diff
@@ -717,6 +776,20 @@ abstract class GeneratedClass {
 }
 
 /**
+ * A wrapper for the source code to be compiled by [[CodeGenerator]].
+ */
+class CodeAndComment(val body: String, val comment: collection.Map[String, String])
+  extends Serializable {
+  override def equals(that: Any): Boolean = that match {
+    case t: CodeAndComment if t eq null => false
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: t will not be null if matched

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 9]
```diff
@@ -706,6 +711,60 @@ class CodegenContext {
...
+   */
+  def getPlaceHolderToComments(): collection.Map[String, String] = {
+    placeHolderToComments
+  }
+
+  /**
+   * Returns the string representation of this expression that is safe to be put in
+   * code comments of generated code. The length is capped at 128 characters.
+   */
+  private[this] def toCommentSafeString(str: String): String = {
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: This is only used by CodeFormatter, I'd like to move it there.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 10]
```diff
@@ -706,6 +711,60 @@ class CodegenContext {
...
+    // they are put in code comment to avoid codegen breaking.
+    // To escape them, single "\" should be prepended to a series of "\" just before "u"
+    // only when the number of "\" is odd.
+    // For example, \u0022 should become to \\u0022
+    // but \\u0022 should not become to \\\u0022 because the first backslash escapes the second one,
+    // and \u0022 will remain, means not escaped.
+    // Otherwise, the runtime Java compiler will fail to compile or code injection can be allowed.
+    // For details, see SPARK-15165.
+    str.substring(0, len).replace("*/", "*\\/")
+      .replaceAll("(^|[^\\\\])(\\\\(\\\\\\\\)*u)", "$1\\\\$2") + suffix
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Maybe we don't need this anymore.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 11]
```diff
@@ -740,6 +813,9 @@ abstract class CodeGenerator[InType <: AnyRef, OutType <: AnyRef] extends Loggin
   /** Binds an input expression to a given input schema */
   protected def bind(in: InType, inputSchema: Seq[Attribute]): InType
 
+  /** Generate the code to be compiled represented as string */
+  def generateCodeAndComments(ctx: CodegenContext, in: InType): CodeAndComment
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Why added this? It seems to me that this required more unnecessary changes.

**Comment by davies** [comment]
@sarutak I like this idea, could you simplify it and minimize the changes?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 11 above)
Comment (sarutak) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: I added this method for testability. We can test whether place holder is put in generated code as we intended.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 8 above)
Comment (sarutak) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Ah, right. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 11 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: We don't need to test it for every generated code, may not worth it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 10 above)
Comment (sarutak) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Yeah, I think we can remove this method rather than move to `CodeFormatter`. What do you think?

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: sgtm

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/GenerateColumnAccessor.scala [hunk 12]
```diff
@@ -224,7 +224,9 @@ object GenerateColumnAccessor extends CodeGenerator[Seq[DataType], ColumnarItera
         }
       }"""
 
-    logDebug(s"Generated ColumnarIterator:\n${CodeFormatter.format(code)}")
+    val code = new CodeAndComment(codeBody, ctx.getPlaceHolderToComments())
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/GenerateColumnAccessor.scala]:
```
val code = new CodeAndComment(codeBody, ctx.getPlaceHolderToComments())
logDebug(s"Generated ColumnarIterator:\n${CodeFormatter.format(code)}")
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 13]
```diff
@@ -706,6 +711,35 @@ class CodegenContext {
...
+
+  /**
+   * get a map of the pair of a place holder and a corresponding comment
+   */
+  def getPlaceHolderToComments(): collection.Map[String, String] = placeHolderToComments
+
+  /**
+   * Register a multi-line comment and return the corresponding place holder
+   */
+  def registerMultilineComment(text: String): String = {
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: private

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 14]
```diff
@@ -706,6 +711,35 @@ class CodegenContext {
...
+  }
+
+  /**
+   * Register a comment and return the corresponding place holder
+   */
+  def registerComment(text: String): String = {
+    if (text.contains("\n") || text.contains("\r")) {
+      registerMultilineComment(text)
+    } else {
+      val placeHolder = s"/*${freshName("comment_placeholder")}*/"
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: /*comment_placeholder*/ -> /*c*/

This is read by human, could be shorter.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 15]
```diff
@@ -717,6 +751,20 @@ abstract class GeneratedClass {
 }
 
 /**
+ * A wrapper for the source code to be compiled by [[CodeGenerator]].
+ */
+class CodeAndComment(val body: String, val comment: collection.Map[String, String])
+  extends Serializable {
+  override def equals(that: Any): Boolean = that match {
+    case null => false
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: this case could be removed

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala [hunk 16]
```diff
@@ -78,8 +77,9 @@ trait CodegenSupport extends SparkPlan {
   final def produce(ctx: CodegenContext, parent: CodegenSupport): String = executeQuery {
     this.parent = parent
     ctx.freshNamePrefix = variablePrefix
+    val placeHolder = ctx.registerComment(s"PRODUCE: ${this.simpleString}")
     s"""
-       |/*** PRODUCE: ${toCommentSafeString(this.simpleString)} */
+       |$placeHolder
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala]: inline `ctx.registerComment(s"PRODUCE: ${this.simpleString}")` here

**Comment by sarutak** [comment]
retest this please.

**Comment by davies** [comment]
LGTM, 
Merging this into master and 2.0, thanks!

**Comment by davies** [comment]
@sarutak Could you create another PR for 1.6? (If we have not fix the security bug in 1.6)

**Comment by sarutak** [comment]
O.K, I'll do it. Need another PR for `branch-1.5` too?

**Comment by davies** [comment]
Maybe 1.5 and 1.6 could share the same PR (if no much conflicts)
